### PR TITLE
Fix reject unknown fields with field label modifier

### LINKED ISSUE #773: `rejectUnknownFields` does not respect `fieldLabelModifier`
Assuming a `FromJSON` instance like this:

```haskell
-- have not run this version, just for illustration
data Foo = Foo
  { a :: Int
  }

instance FromJSON Foo where
  parseJSON = genericParseJSON defaultOptions { rejectUnknownFields = True, fieldLabelModifier = rename }
    where
      rename   "a" = "b"
      rename other = other
```

Parsing of will fail with a valid input like: `{"b":1}` will fail with an error like:

```
parsing Foo(Foo) failed, unknown fields ["b"]
```

The `fieldLabelModifier` in the code path to `knownFields` does never apply the `fieldLabelModifier`. 

I'm trying to come up with a PR to fix this.

### PR DESCRIPTION
**Pull request by mbj**
My attempt to fix #773 I reported earlier.

This PR comes in 2 commits. The first adds error message gets for the TH parsing. The 2nd commit fixes the problem I reported with changing the (IMO wrong) error expectations to demonstrate the bugfix.

There are some minor implementation trade-offs. I'll comment in code below to highlight my thoughts on them.

### CODE REVIEW

File: Data/Aeson/Types/FromJSON.hs [hunk 1]
```diff
@@ -1260,15 +1260,15 @@ instance (ProductFromJSON arity f, ProductSize f
 --------------------------------------------------------------------------------
 
 class FieldNames f where
-    fieldNames :: f a -> [Text] -> [Text]
+    fieldNames :: f a -> [String] -> [String]
```
Comment (mbj) [inline Data/Aeson/Types/FromJSON.hs:1263]: I changed the type signature from `[Text]` to `[String]` deliberately to be able to call the `fieldLabelModifier` without a `pack` / `unpack` cycle.

File: Data/Aeson/Types/FromJSON.hs [hunk 2]
```diff
@@ -1282,9 +1282,10 @@ instance ( FieldNames f
         \obj -> checkUnknown obj >> recordParseJSON' p obj
         where
             knownFields :: H.HashMap Text ()
-            knownFields = H.fromList $ map (,()) $
-                fieldNames (undefined :: f a)
-                [pack (tagFieldName (sumEncoding opts)) | fromTaggedSum]
+            knownFields = H.fromList $ map ((,()) . pack) $
+                [(tagFieldName (sumEncoding opts)) | fromTaggedSum] <>
+                (fieldLabelModifier opts <$> fieldNames (undefined :: f a) [])
```
Comment (mbj) [inline Data/Aeson/Types/FromJSON.hs:1287]: The `fieldLabelModifier` should not be applied to the "tag" fields. Hence I could not anymore pass the tag fields into the field names reflection.

I had to choose between: A list concatentation (and these can be inefficient) and a union of 2 hashes, one build from the tag fields and one from the reflected field names.

I went with the list concatenation, as I assume that the number of tag fields will usually be lower and appending to a "small" list is quite fast. Hence also putting the "smaller" list on the LHs of `<>`.

File: changelog.md [hunk 3]
```diff
@@ -3,6 +3,7 @@ For the latest version of this document, please see [https://github.com/bos/aeso
 #### 1.4.7.1
 
 * GHC 8.10 compatibility, thanks to Ryan Scott.
+* Fix bug in `rejectUnknownFields` not respecting `fieldLabelModifier`.
```
Comment (mbj) [inline changelog.md:6]: I've noted that the changelog entries typically say `thanks to $Author` but I think I should not thank myself, especially while my code is under review. I hope the reviewer edits it in, if appropriate ;)

File: tests/ErrorMessages.hs [hunk 4]
```diff
@@ -141,7 +141,7 @@ outputGeneric choice = concat
       (select
         thSomeTypeParseJSONRejectUnknownFields
         gSomeTypeParseJSONRejectUnknownFields)
-      [ "{\"tag\": \"record\", \"testOne\": 1.0, \"testZero\": 1}"
+      [ "{\"tag\": \"record\", \"testone\": 1.0, \"testZero\": 1}"
```
Comment (mbj) [inline tests/ErrorMessages.hs:144]: `optsRejectUnknownFields` actually had the `fieldLabelModifier = map toLower` set all the time. Hence this test input should have rejected the `testOne` originally and prior to my bugfix did not. 

I chose to not change the expectations in the golden test but instead made the test input comply with the expectation.

File: tests/golden/generic.expected [hunk 5]
```diff
@@ -34,7 +34,7 @@ Error in $: not enough input. Expecting json list value
 SomeType (reject unknown fields)
 Error in $: parsing Types.SomeType(Record) failed, unknown fields: ["testZero"]
 Error in $: parsing Types.SomeType failed, expected Object with key "tag" containing one of ["nullary","unary","product","record","list"], key "tag" not found
-Error in $: parsing Types.SomeType(Record) failed, unknown fields: ["testtwo","testone","testthree"]
+Error in $.testone: parsing Double failed, unexpected Boolean
```
Comment (mbj) [inline tests/golden/generic.expected:37]: The "real" error message was uncovered by my change. The previously reported fields where not unknown as the test case has `fieldLabelModifier = map toLower` applied.

File: tests/golden/th.expected [hunk 6]
```diff
@@ -34,7 +34,7 @@ Error in $: not enough input. Expecting json list value
 SomeType (reject unknown fields)
 Error in $: Unknown fields: ["testZero"]
 Error in $: key "tag" not found
-Error in $: Unknown fields: ["testtwo","testone","testthree"]
+Error in $.testone: parsing Double failed, unexpected Boolean
```
Comment (mbj) [inline tests/golden/th.expected:37]: Same as https://github.com/bos/aeson/pull/774/commits/70d0927aa32481d5f4d312c8c550763fe7d41e27#r419038724 but for TH.

I noticed there where no tests to update to validate the TH bugfix so I created the first commit and extracted it so the reviewer can also see the TH behavior change being reflected in tests.

Comment (mbj) [reply tests/golden/th.expected:37]: For this comment to make sense, look at the 2nd commit in isolation.

File: Data/Aeson/Types/FromJSON.hs (hunk 2 above)
Comment (bergmark) [reply Data/Aeson/Types/FromJSON.hs:1287]: Sounds good!

File: changelog.md (hunk 3 above)
Comment (bergmark) [reply changelog.md:6]: Yeah i always go through this before release, but 1.4.7.1 is already out. We usually put new stuff under an Upcoming header, but I can do that when merging.

**Comment by mbj** [comment]
@bergmark Thanks for the quick review. What is the process after your approval?

**Comment by bergmark** [comment]
Nothing for this PR!

@phadej Did you have any other breaking changes you wanted to include in the next release or should we ship 1.5.0.0?

**Comment by mbj** [comment]
> Nothing for this PR!
> 
> @phadej Did you have any other breaking changes you wanted to include in the next release or should we ship 1.5.0.0?

Any chance we can get this released? Can I help somewhere to make it happen?

**Comment by phadej** [comment]
@bergmark the scanner rework is something which will force breaking change. Yet, I don't think that will be ready soon.

I'd merge
- #653
and fix #700

and then I think 1.5 would be solid.

**Comment by mbj** [comment]
> and then I think 1.5 would be solid.

Chances we can get 1.4.7.2 meanwhile, which includes this PR? The [changelog](https://github.com/bos/aeson/blob/master/changelog.md) so far does not list a breaking change, so cutting another 1.4 may be eligible. This way we do not have to wait to get the fix out.

**Comment by bergmark** [comment]
Sorry for the delay @mbj, 1.5.0.0 was released now. We already had some breaking changes that weren't listed in the changelog. Best place to check for things like that is versioning labels on the current milestone, e.g. https://github.com/bos/aeson/milestone/21?closed=1

**Comment by mbj** [comment]
@bergmark Thanks for the release and pinging me, much appreciated. Cheers.

**Pull request merged**